contributors/design-proposals/storage/local-storage-pv.md

```diff
@@ -761,3 +761,239 @@ The central manager will poll each worker for their discovered volumes and creat
+
+New field will be added into the configuration ConfigMap, to specify available
+volume group for lvm, the group should be created by the admin in advance:
+
+```
+kind: ConfigMap
+metadata:
+  name: local-volume-config
+  namespace: kube-system
+data:
+ storageClassMap: |
+    local-static:
+      hostDir: "/mnt/ssds"
+      mountDir: "/local-ssds"
+    local-dynamic:
+      lvm:
+        volumeGroup: "volumeGroup1"
+      blockCleanerCommand: "./quick_reset.sh"
+```
+
+A StorageManager instance is responsible for:
+
+1. Monitor available capacity for provisioning, and update the capacity in StorageClass if
+there is a change.
+2. Implement interface `CreateLocalVolume(claim *v1.PersistentVolumeClaim) error`
+to create PVs and the local volumes behind them.
+3. Implement interface `DeleteLocalVolume(pv *v1.PersistentVolume) error` to
+clean up the local volume behind a PV.
+
+##### Reconcile Capacity
+
+The StorageManager is expected to reconcile capacity of StorageClasses in its local
+`storageClassMap` regularly. Particularly, for LVM, the proposed workflow is:
+
+1. Walks through classes in its `storageClassMap` regularly
+2. Fetch the VG of each class if `volumeGroup` is specified
+3. Compare the value in `VG size` with the capacity in StorageClass API object
+4. Update the capacity in StorageClass object if there is a difference
+
+**Note:** For the alpha phase, thinly-provisioned VG (overcommit) is not supported,
+We'll consider it in a subsequent phase.
+
+##### Change in Populator
+
+In addition to PV, the `populator` will also be extended to watch PVC objects of all namespaces.
+When the `populator` finds:
+
+1. An unbound PVC is annotated with `annSelectedNode`, and its node name match the value
+of the annotation.
+2. Value of the `annStorageProvisioner` on the PVC matches the name of the provisioner.
+
+It then trigger volume provsioning, detailed in the volume creation section.
+
+For these purposes, the following permissions are required:
+
+* list/get PersistentVolumeClaims
+* list/get/update StorageClasses
+
+##### Volume Creation
+
+During start-up of the provisioner, a new work queue `ProvisionQueue` will be created,
+and passed into both `populator` and `storageManager`:
+
+```
+ProvisionQueue *workqueue.Type
+```
+When it finds a PVC is in need of privsioning, as desicribed above,
+the `populator` will pass it into the queue to trigger volume privisioning;
+`StorageManager` will be the consumer, it will watch the queue and
+call its `CreateLocalVolume` function to:
+
+1. Create volumes behind a PV accordingly
+2. Create a PV object that is pre-bound to the PVC
+
+The proposed workflow is:
+
+1. Check if there is already a PV pre-bound to the PVC, if so, return directly
+(This can happen when a volume has already been provisioned for a PVC,
+but has not been bound to the PVC by the controller yet).
+2. Check if `annProvisionerTopology` is set on the PVC. If not, then it means

```

Comment:
I think this could be done one level up

Reply:
you mean in the populator?
In the populator we only filter incoming pvcs, but for this case, we'll need to treat it as "failure" and signal back to the scheduler to retry dynamic provisioning by removing the "annSelectedNode" annotation of the PVC. So I think it should be better to place it here in the manager.

Reply:
I mean in the layer that is managing whether or not to call CreateVolume(). The main question I ask myself for every operation is "Is this something that is specific to the LVM implementation?"

Reply:
Have clarified to @msau42, `CreateVolume` here is a generic interface and not specific to the LVM.